packages/commonwealth/server/migrations/20231005213247-assign_master_profile.js

```diff
@@ -0,0 +1,76 @@
+'use strict';
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.sequelize.transaction(async (t) => {
+      await queryInterface.sequelize.query(
+        `
+        ALTER TABLE "Addresses" ADD COLUMN IF NOT EXISTS "legacy_user_id" INTEGER NULL;
+        ALTER TABLE "Addresses" ADD COLUMN IF NOT EXISTS "legacy_profile_id" INTEGER NULL;
+        `,
+        { raw: true, transaction: t }
+      );
+    });
+
+    await queryInterface.sequelize.transaction(async (t) => {
+      try {
+        await queryInterface.sequelize.query(
+          `
+          WITH MasterProfiles AS (
+            SELECT
+              A.hex,
+              P.id AS master_profile_id,
+              P.user_id AS master_user_id
+            FROM "Addresses" A
+            LEFT JOIN (
+              SELECT
+                A2.hex,
+                P2.id,
+                P2.user_id,
+                RANK() OVER (PARTITION BY A2.hex ORDER BY
+                  COUNT(CASE WHEN P2.profile_name IS NOT NULL THEN 1 ELSE NULL END) +
+                  COUNT(CASE WHEN P2.email IS NOT NULL THEN 1 ELSE NULL END) +
+                  COUNT(CASE WHEN P2.bio IS NOT NULL THEN 1 ELSE NULL END) +
+                  COUNT(CASE WHEN P2.avatar_url IS NOT NULL THEN 1 ELSE NULL END) +
+                  COUNT(CASE WHEN P2.socials IS NOT NULL THEN 1 ELSE NULL END) +
+                  COUNT(CASE WHEN P2.background_image IS NOT NULL THEN 1 ELSE NULL END) DESC
+                ) AS rank
+              FROM "Addresses" A2
+              LEFT JOIN "Profiles" P2 ON A2.profile_id = P2.id
+              WHERE A2.hex IS NOT NULL
+              AND P2.id IS NOT NULL
+              GROUP BY A2.hex, P2.id, P2.user_id
+            ) AS P ON A.hex = P.hex
+            WHERE P.rank = 1
+          )
+          UPDATE "Addresses" A
+          SET
+            legacy_user_id = A.user_id,
+            legacy_profile_id = A.profile_id,
+            user_id = MP.master_user_id,
+            profile_id = MP.master_profile_id
+          FROM MasterProfiles MP
+          WHERE A.hex = MP.hex;
+          `,
+          { transaction: t }
+        );
+      } catch (e) {
+        // console.log('error', e);
+        throw new Error(e);
+      }
+    });
```

Comment:
@timolegros This is a working example of using the Profile with most-populated fields.

I have been trying your idea of making a new profile that combines populated fields from existing profiles, favoring the latest update. But so far I have failed due to the complexity. Do you feel like taking a crack at it or finding time to pair on it? 

We can also talk about it in OH.

Reply:
Let's discuss in OH and pair after if needed.

Reply:
To note though you don't have to do absolutely everything in a single step. You shouldn't make queries in a JS loop but that doesn't mean you can't pull all the profile data into JS, consolidate the data (create the master profile objects), then submit a single bulkCreate operation to create all the master profiles, and then finally link the master profiles to addresses.

Reply:
Can you do a bulkCreate in a migration? I don't see any precedent.

Reply:
nvm I see bulkInsert